plugin/semisync/semisync_master.cc

```diff
@@ -513,13 +523,29 @@ void ReplSemiSyncMaster::lock() { mysql_mutex_lock(&LOCK_binlog_); }
 
 void ReplSemiSyncMaster::unlock() { mysql_mutex_unlock(&LOCK_binlog_); }
 
-void ReplSemiSyncMaster::add_slave() {
+void ReplSemiSyncMaster::add_slave(int server_id) {
   lock();
   rpl_semi_sync_master_clients++;
+  rpl_semi_sync_master_whitelist_set_lock.lock();
```

Comment:
Use `std::lock_guard` here as well, probably wrap into a new code block `{...}`
```suggestion
  std::lock_guard<std::mutex> guard(rpl_semi_sync_master_whitelist_set_lock);
```